### DISCUSSION TITLE
fatal error: 'utf-8' codec can't encode characters in position

### DISCUSSION THREAD

**Discussion by Russsgithub** [savonet/liquidsoap#1863] [Installation]
I'm doing this 
```
s = single("s3://S3Bucket/1630132773673_Netherlands_Törf_Bistoe_nog_in_leven_2017.mp3")
output.icecast(description="audio", icy_metadata="true", format="audio/aac", %fdkaac(channels=2, samplerate=48000, bitrate=320, afterburner=true, transmux="adts", aot="mpeg4_aac_lc", sbr_mode=false), host="ice", port=8000, password="pass",mount="test",mksafe(clock(s)))
```

but get this error
```
fatal error: 'utf-8' codec can't encode characters in position 27-28: surrogates not allowed
```
the offending character ```ö``` should get encoded as expected ?

locale returns
```
LANG=en_GB.UTF-8
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE=C
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
```

I have the latest git main code installed

**Comment by snahor** [comment]
Try setting `LC_ALL`.

**Comment by millicent-audio-mirror** [comment]
Thanks @snahor , 
I don't know a lot about locale , but from a little research it seems to me that LC_ALL overrides the other settings , so if LANG is set to en_GB-utf-8 and I set LC_ALL to the same it will override and I'm still in the same place I was.

Saying that I will give it a go and see.

**Comment by Russsgithub** [comment]
@snahor , @millicent-audio-mirror is my other account, i didn't noticed i was logged in to that account for last reply.

@toots says here https://github.com/savonet/liquidsoap/issues/697#issuecomment-458276121 that liq gets its language settings from LANG, which is set.

I have tried setting LC_ALL, but it does not effect the result.

Thanks

**Comment by toots** [comment]
Hi @Russsgithub ! Could you send me a copy of this file? I believe this error either comes from your system or from the camomile lib but not from us.

**Comment by Russsgithub** [comment]
There's a copy of the file here

https://drive.google.com/file/d/1cmfHk9ReWKauIJQwd6iy1oPCoQFdg5N-/view?usp=drivesdk

If I run the was s3 cp command that comes up in the error awscli downloads the file as expected.

**Comment by toots** [comment]
I have downloaded the file. I can confirm the presence of the non-ascii character `ö` in both the filename and metadata but it seems to be working fine here. I'd suggest that this is an issue in your local configuration. Moving this to a discussion, please advise otherwise. Hopefully you find a way to fix it on your end.

**Comment by Russsgithub** [comment]
@toots 
The file works from a local directory
```
'annotate:title="Title 1",artist="Artist 1":/local_folder/1630132773673_Netherlands_Törf_Bistoe_nog_in_leven_2017.mp3'
```

but if i try the remote file (s3)
```
'annotate:title="Title 1",artist="Artist 1":s3://s3bucket/1630132773673_Netherlands_Törf_Bistoe_nog_in_leven_2017.mp3'
```
i get
```
fatal error: 'utf-8' codec can't encode characters in position 27-28: surrogates not allowed

```
but the file will download from the command line with the awscli tool.

If liquidsoap can read the file name properly and play it from a local folder, and awscli can read the remote file is this not  coming from how liquidsoap is talking to awscli ??

**Comment by Russsgithub** [comment]
I forgot to mention that this is a recent thing. Files with such characters have played fine up to maybe a week or two ago.

This doesn't discount the possibility of issues on my system. Though the error occurs on both my production machine and my dev machine.

I'm a bit flummoxed. ( I like that word! )

**Reply by smimram** [reply]
There was quite some work from @toots recently on charset encoding which might explain this: 2b9e35005ee7ed7d50edd21e8f1a26e79d03f787

**Comment by Russsgithub** [comment]
This actually seems to be that I updated the awscli from v1 to v2 . 

v2 is not allowing those character from liquidsoap, but is from the command line. I find this really strange as it doesn't appear to me that liq changes the string at all , and as i mentioned above , if i copy the command given in the error it works from the command line ....

I've just downgraded to awscli v1 and liq can download the file fine.

I'll drop an issue at the awscliv2 git page.

**Comment by toots** [comment]
Okay y'all thanks for insisting a bit tho, it looks like there's also something we should do on our own. Here's a relevant stackoverflow: https://stackoverflow.com/questions/27366479/python-3-os-walk-file-paths-unicodeencodeerror-utf-8-codec-cant-encode-s

I recently switched all string escaping for logging purposes to log strings as utf8 but did not realize that filenames could in fact not be utf8 strings if your system or filesystem is not setup with utf8 internally.

Will push something shortly.

**Comment by toots** [comment]
https://github.com/savonet/liquidsoap/commit/3ef88bbbba8557668956627fca55ab3cc41de63d should help. @Russsgithub any chance you could test again with this commit?

**Comment by Russsgithub** [comment]
I can i just have to reinstall awscli v2.

**Comment by Russsgithub** [comment]
I'm afraid not , still getting the utf-8 error on awscliv2

Still works on awscliv1

**Comment by toots** [comment]
Ok thanks for checking. I'm suspecting an error with awscli then since I believe it is indeed written in python and the error corresponds to the stackoverflow message I linked above. Thanks for checking!

**Reply by Russsgithub** [reply]
No worries, I haven't reported it over at awscliv2 git issues. I don't quite know how to word it , as I don't really understand what liq is doing to the url string that it uses.

**Comment by toots** [comment]
I would describe what you are trying to do, link to the s3 file and stackoverflow message. On the liquidsoap side, we are executing this command line:
```
aws s3 cp "s3://<bucket>/<filename>" "<local filename>"
```
We are escaping the utf8 file name following the patterns described in this PR: https://github.com/savonet/liquidsoap/pull/1839

**Comment by toots** [comment]
Let me try one more thing tho!

**Comment by toots** [comment]
Yeah, no, feel free to submit a but to aws cli 🙂

**Comment by Russsgithub** [comment]
:), Will do.

**Comment by Russsgithub** [comment]
@toots yesterdays change to s3 string parsing is causing me issues.

I have been passing s3 urls with via annotate enclosed in ```"""``` to deal with escaping.

The code changes you made yesterday stopped this from working, but when removing the ```"""``` the string parses fine.

Now I get ```Failed to resolve``` error if characters such as ```'``` are in the url.

I've tried passing the url's inside quotes, which is not accepted. and escaping the characters, which also doesn't work

**Comment by Russsgithub** [comment]
```
"annotate:duration='243.77',liq_on_offset='236.40235827664398',lastPlayed='2021-08-30T09:08:16.000Z',liq_cue_in='0',track_peak='0.787910',intro='music',outro='noise',style='Ambient',mix_type='Mix',loudness='5',file_id='5fc923500eaa872e2cdc7c26',liq_amplify='-5.600000dB',type='music',title='Haxaal's Dream',artist='Yage':s3://S3BUCKET/1607017292310_17-yage-haxaals_dream.mp3"
```
This is an example of a failing annotate string. The title contains ```'``` . So it isn't just the url that is effected.

**Comment by Russsgithub** [comment]
```
Error while parsing annotate URI "duration='309.19',liq_cue_in='0',
loudness='7',file_id='5fb6d3340eaa872e2cdc5663',liq_amplify='+2.170000dB',type='music',title='Theo's
 Solar System',artist='b p r':s3://S3BUCKET/1605817135557_Theo's_Solar_System.mp3": Char 132-1
33: Syntax error
```

**Comment by Russsgithub** [comment]
I think s3 url's work. But annotated titles with escapeable characters e.g. ```'``` need escaping in annotate i.e (in title).
```
"annotate:duration='243.77',liq_on_offset='236.40235827664398',lastPlayed='2021-08-30T09:08:16.000Z',liq_cue_in='0',track_peak='0.787910',intro='music',outro='noise',style='Ambient',mix_type='Mix',loudness='5',file_id='5fc923500eaa872e2cdc7c26',liq_amplify='-5.600000dB',type='music',title='Haxaal\\'s Dream',artist='Yage':s3://S3BUCKET/1607017292310_17-yage-haxaals_dream.mp3"
```

**Comment by toots** [comment]
Yes, that's expected, otherwise there's no way to know if that `'` is part of the title string or the end of the quotation string.

**Comment by Russsgithub** [comment]
Great. Thanks.

On 1 Sep 2021 13:03, Romain Beauxis ***@***.***> wrote:

Yes, that's expected, otherwise there's no way to know if that ' is part of the title string or the end of the quotation string.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/savonet/liquidsoap/discussions/1863#discussioncomment-1264880>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB627QKLX2MHPOASWGWSNMDT7YJBBANCNFSM5DAUDSKA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
